### PR TITLE
Cloudflare sync template

### PR DESCRIPTION
**Pull request by SomeHats**
Adds a template for tldraw+sync on cloudflare workers. The code is mostly adapted from the demo server itself (with some fixes/tweaks backported across). I've added a bunch of comments, and some guidance in 
 the README - [rendered version here](https://github.com/tldraw/tldraw/blob/alex/sync-workers-template/templates/cloudflare-workers/README.md). Let me know what you think!

### Change type

- [x] `other`

### CODE REVIEW

File: templates/cloudflare-workers/client/App.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	})
+
+	return (
+		<div style={{ position: 'fixed', inset: 0 }}>
+			<Tldraw
+				// we can pass the connected store into the Tldraw component which will handle
+				// loading states & enable multiplayer UX like cursors & a presence menu
+				store={store}
+				onMount={(editor) => {
+					// when the editor is ready, we need to register out bookmark unfurling service
```
Comment (ds300) [inline templates/cloudflare-workers/client/App.tsx:34]:
```suggestion
					// when the editor is ready, we need to register our bookmark unfurling service
```

File: templates/cloudflare-workers/worker/TldrawDurableObject.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+				// create a new TLSocketRoom. This handles all the sync protocol & websocket connections.
+				// it's up to us to persist the room state to R2 when needed though.
+				return new TLSocketRoom<TLRecord, void>({
+					schema,
+					initialSnapshot,
+					onSessionRemoved: async (_room, args) => {
+						// persist whenever the number of active sessions drops to zero
+						if (args.numSessionsRemaining > 0) return
+						await this.schedulePersistToR2()
+					},
```
Comment (ds300) [inline templates/cloudflare-workers/worker/TldrawDurableObject.ts:106]: I don't think this is actually necessary. onDataChange will have been called if the room data changed so there's no need to do it again here.

```suggestion
```

File: templates/cloudflare-workers/worker/TldrawDurableObject.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		const clock = room.getCurrentDocumentClock()
+		if (this.lastPersistedClock === clock) return
+
+		// convert the room to JSON and upload it to R2
+		const snapshot = JSON.stringify(room.getCurrentSnapshot())
+		await this.r2.put(`rooms/${this.roomId}`, snapshot)
+
+		this.lastPersistedClock = clock
+	}, 10_000)
+	lastPersistedClock = 0
```
Comment (ds300) [inline templates/cloudflare-workers/worker/TldrawDurableObject.ts:133]: I think we can get rid of `lastPersistedClock` too. It's a very minor optimization and not worth polluting this template with, especially if we remove the sessionRemoved hanlder

File: templates/cloudflare-workers/worker/assetUploads.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		let end
+		if ('suffix' in object.range) {
+			start = object.size - object.range.suffix
+			end = object.size - 1
+		} else {
+			start = object.range.offset ?? 0
+			end = object.range.length ? start + object.range.length - 1 : object.size - 1
+		}
+		headers.set('content-range', `bytes ${start}-${end}/${object.size}`)
+	}
```
Comment (ds300) [inline templates/cloudflare-workers/worker/assetUploads.ts:68]: it's surprising that we need to do this manually, might be worth a comment.

File: templates/cloudflare-workers/worker/bookmarkUnfurling.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	try {
+		await new HTMLRewriter()
+			.on('meta', meta$)
+			.on('title', title$)
+			.on('link', icon$)
+			.transform((await fetch(url)) as any)
+			.blob()
+	} catch {
+		return null
+	}
```
Comment (ds300) [inline templates/cloudflare-workers/worker/bookmarkUnfurling.ts:25]: (followup work) I want us to put this whole thing in a npm package. Maybe even in a separate repo under MIT. It's definitely an unfulfilled need, and people will probably make PRs with improvements for different edge cases.

File: templates/cloudflare-workers/worker/worker.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+export { TldrawDurableObject } from './TldrawDurableObject'
+
+// we use itty-router (https://itty.dev/) to handle routing. in this example we turn on CORS because
+// we're hosting the worker separately to the client. you should restrict this to your own domain.
+const { preflight, corsify } = cors({ origin: '*' })
+const router = AutoRouter<IRequest, [env: Environment, ctx: ExecutionContext]>({
+	before: [preflight],
+	finally: [corsify],
+	catch: (e) => {
+		console.log(e)
```
Comment (ds300) [inline templates/cloudflare-workers/worker/worker.ts:16]:
```suggestion
		console.error(e)
```

File: templates/cloudflare-workers/worker/worker.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import { handleAssetDownload, handleAssetUpload } from './assetUploads'
+import { handleUnfurlUrl } from './bookmarkUnfurling'
+import { Environment } from './types'
+
+// make sure our sync durable object is made available to cloudflare
+export { TldrawDurableObject } from './TldrawDurableObject'
+
+// we use itty-router (https://itty.dev/) to handle routing. in this example we turn on CORS because
+// we're hosting the worker separately to the client. you should restrict this to your own domain.
+const { preflight, corsify } = cors({ origin: '*' })
```
Comment (ds300) [inline templates/cloudflare-workers/worker/worker.ts:11]:
```suggestion
// we're hosting the worker separately to the client. you should restrict this to your own domains.
const { preflight, corsify } = cors({ origin: ['*'] })
```

might be a tiny bit easier for folks to adapt if it's already an array.

**Review by ds300 — APPROVED** [review]
This is excellent work. Love the readme, love all the comments. It's a really compelling story. Can't wait to put it in people's hands.

**Review by mimecuvalo — COMMENTED** [review]
just popping in to say: wow.
🤩

File: templates/cloudflare-workers/README.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+To deploy this example, you'll need to create a cloudflare account and create an R2 bucket to store
+your data. Update `bucket_name = 'tldraw-content'` in [`wrangler.toml`](./wrangler.toml) with the
+name of your new bucket.
+
+Run `wrangler deploy` to deploy your backend. This should give you a workers.dev URL, but you can
+also [configure a custom
+domain](https://developers.cloudflare.com/workers/configuration/routing/custom-domains/).
+
+Finally, deploy your client HTML & JavaScript. Create a production build with
+`TLDRAW_WORKER_URL=your.workers.domain.com yarn build`. Publish the resulting build (in `dist/`) on
```
Comment (ds300) [inline templates/cloudflare-workers/README.md:70]:
```suggestion
`TLDRAW_WORKER_URL=https://your.workers.domain.com yarn build`. Publish the resulting build (in `dist/`) on
```
this won't work without the protocol

File: templates/cloudflare-workers/vite.config.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+export default defineConfig(({ mode }) => {
+	if (mode === 'production' && !process.env.TLDRAW_WORKER_URL) {
+		throw new Error('TLDRAW_WORKER_URL must be set in production')
+	}
+
+	return {
+		plugins: [react()],
+		define: {
+			'process.env.TLDRAW_WORKER_URL':
+				process.env.TLDRAW_WORKER_URL ?? '`http://${location.hostname}:5172`',
```
Comment (ds300) [inline templates/cloudflare-workers/vite.config.ts:14]:
```suggestion
				process.env.TLDRAW_WORKER_URL ?? '`http://${location.host}`',
```
i think we can do this and avoid hardcoding the port in too many places

File: templates/cloudflare-workers/worker/worker.ts (hunk 7 above)
Comment (ds300) [reply templates/cloudflare-workers/worker/worker.ts:11]: errr nvm it's not the same behaviour in an array

File: templates/cloudflare-workers/vite.config.ts (hunk 9 above)
Comment (SomeHats) [reply templates/cloudflare-workers/vite.config.ts:14]: the port here is different: it's 5173 for the client, and 5172 for the server

Comment (ds300) [reply templates/cloudflare-workers/vite.config.ts:14]: silly me

**Pull request merged**